modin/backends/pandas/query_compiler.py

```diff
@@ -1113,39 +1122,54 @@ def map_func(df):
         )
         result = self.__constructor__(new_modin_frame)
 
-        if is_all_multi_list:
+        def compute_index(index, columns, consider_index=True, consider_columns=True):
+            def compute_axis(axis):
+                return [
+                    axis.get_level_values(lvl).unique()
+                    for lvl in np.arange(axis.nlevels)
+                ]
+
+            new_index = compute_axis(index) if consider_index else index
+            new_columns = compute_axis(columns) if consider_columns else columns
+            return pandas.MultiIndex.from_product([*new_columns, *new_index])
+
+        if is_all_multi_list and is_correct_len:
             result = result.sort_index()
             index_level_values = [lvl for lvl in obj.index.levels]
-            columns_level_values = [
-                real_cols_bkp.get_level_values(lvl).unique()
-                for lvl in np.arange(real_cols_bkp.nlevels)
-            ]
-            result.index = pandas.MultiIndex.from_product(
-                [*columns_level_values, *index_level_values]
+
+            result.index = compute_index(
+                index_level_values, real_cols_bkp, consider_index=False
             )
             return result
 
         if need_reindex:
-            if isinstance(self.index, pandas.MultiIndex):
-                index_level_values = [
-                    self.index.get_level_values(lvl).unique()
-                    for lvl in np.arange(self.index.nlevels)
-                ]
-                new_index = pandas.MultiIndex.from_product(
-                    [self.columns, *index_level_values]
-                )
+            if is_correct_len:
+                if isinstance(self.index, pandas.MultiIndex):
+                    new_index = compute_index(
+                        self.index, [self.columns], consider_columns=False
+                    )
+                else:
+                    if isinstance(self.columns, pandas.MultiIndex):
+                        new_index = compute_index(
+                            [self.index], self.columns, consider_index=False
+                        )
+                    else:
+                        new_index = compute_index(
+                            [self.index],
+                            [self.columns],
+                            consider_index=False,
+                            consider_columns=False,
+                        )
```

Comment:
If we will move the logic of wrapping or not wrapping indices in a list, then we can replace this if-else tree in these three lines:
```python
should_recompute_index = isinstance(self.index, pandas.MultiIndex)
should_recompute_columns = not should_recompute_index and isinstance(self.columns, pandas.MultiIndex)

new_index = compute_index(self.index, self.columns, should_recompute_index, should_recompute_columns)
```

Reply:
fixed